debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/infinispan/InfinispanLogMinerTransactionCache.java

```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Override
+    public boolean containsTransaction(String transactionId) {
+        return eventIdsByTransactionId.containsKey(transactionId);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return eventIdsByTransactionId.isEmpty();
+    }
+
+    @Override
+    public int getTransactionCount() {
+        return eventIdsByTransactionId.size();
+    }
+
+    @Override
+    public <R> R streamTransactionsAndReturn(Function<Stream<InfinispanTransaction>, R> consumer) {
+        waitForAllWrites();
+        try (Stream<InfinispanTransaction> stream = transactionCache.values().stream()) {
+            return consumer.apply(stream);
+        }
+    }
+
+    @Override
+    public void transactions(Consumer<Stream<InfinispanTransaction>> consumer) {
+        waitForAllWrites();
+        try (Stream<InfinispanTransaction> stream = transactionCache.values().stream()) {
+            consumer.accept(stream);
+        }
+    }
+
+    @Override
+    public void eventKeys(Consumer<Stream<String>> consumer) {
+        waitForAllWrites();
+        try (Stream<String> stream = eventCache.keySet().stream()) {
+            consumer.accept(stream);
+        }
+    }
+
+    @Override
+    public Iterator<LogMinerEvent> eventsIterator(InfinispanTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId())
+                .stream()
+                .map(eventId -> getTransactionEvent(transaction, eventId))
+                .iterator();
+    }
+
+    @Override
+    public LogMinerEvent getTransactionEvent(InfinispanTransaction transaction, int eventKey) {
+        waitForAllWrites();
+        return eventCache.get(transaction.getEventId(eventKey));
+    }
+
+    @Override
+    public InfinispanTransaction getAndRemoveTransaction(String transactionId) {
+        waitForAllWrites();
+        // Intentionally blocking
+        return transactionCache.remove(transactionId);
+    }
+
+    @Override
+    public void addTransactionEvent(InfinispanTransaction transaction, int eventKey, LogMinerEvent event) {
+        putAsync(eventCache, transaction.getEventId(eventKey), event);
+        eventIdsByTransactionId.get(transaction.getTransactionId()).add(eventKey);
+    }
+
+    @Override
+    public void removeTransactionEvents(InfinispanTransaction transaction) {
+        waitForAllWrites();
+
+        eventIdsByTransactionId.get(transaction.getTransactionId())
+                .descendingSet()
+                .stream()
+                .map(transaction::getEventId)
+                .forEach(eventCache::remove);
+
+        eventIdsByTransactionId.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean removeTransactionEventWithRowId(InfinispanTransaction transaction, String rowId) {
+        waitForAllWrites();
+
+        final TreeSet<Integer> eventIds = eventIdsByTransactionId.get(transaction.getTransactionId());
+        return eventIds.descendingSet().stream()
```

Comment:
copy treeset before streaming.

Reply:
If we think this is an issue, then I'd prefer using an alternative that avoids the copy. This method in particular is one of the critical reasons for this issue, so it needs to be as fast and efficient as possible. For large transactions, a copy of the set is `O(n)` and if I have a large transaction with hundreds of thousands or millions of partial rollbacks, that cost will eat away at any performance gains we're trying to achieve I believe.

Reply:
I think for these specific cases, I'll fall back to a traditional loop to avoid the stream concurrency concerns.

Reply:
I think as implement it's fine since findFirst results in an Optional and stream is no longer an issue.  The issue may just be readability in that at my first glance, I mistook it to be removing things from the stream while it was being processed.